File: packages/awiki_im_core/lib/src/models/group.dart
import 'message.dart';

enum GroupIdentityMode { handle, didOnly }

class GroupSummary {
  const GroupSummary({
    this.id,
    required this.did,
    this.name,
    this.displayName,
    this.avatarUri,
    this.myRole,
    this.membershipStatus,
    this.memberCount,
    this.lastMessageAt,
  });

  final String? id;
  final String did;
  final String? name;
  final String? displayName;
  final String? avatarUri;
  final String? myRole;
  final String? membershipStatus;
  final int? memberCount;
  final String? lastMessageAt;
}

class GroupSnapshot extends GroupSummary {
  const GroupSnapshot({
    super.id,
    required super.did,
    super.name,
    super.displayName,
    super.avatarUri,
    super.myRole,
    this.description,
    super.membershipStatus,
    super.memberCount,
    super.lastMessageAt,
  });

  final String? description;
}

class GroupMember {
  const GroupMember({
    this.membershipId,
    this.peerPersonaId,
    this.did,
    this.credentialDid,
    this.handle,
    this.role,
    this.status,
    this.joinedAt,
    this.subjectType,
  });

  final String? membershipId;
  final String? peerPersonaId;
  final String? did;
  final String? credentialDid;
  final String? handle;
  final String? role;
  final String? status;
  final String? joinedAt;
  final String? subjectType;
}

class GroupDiscoverability {
  const GroupDiscoverability._(this.value);

  static const private = GroupDiscoverability._('private');
  static const public = GroupDiscoverability._('public');
  static const unlisted = GroupDiscoverability._('unlisted');

  factory GroupDiscoverability.custom(String value) {
    final trimmed = value.trim();
    if (trimmed.isEmpty) {
      throw ArgumentError.value(value, 'value', 'must not be empty');
    }
    return GroupDiscoverability._(trimmed);
  }

  final String value;
}

class GroupAdmissionMode {
  const GroupAdmissionMode._(this.value);

  static const openJoin = GroupAdmissionMode._('open-join');
  static const inviteOnly = GroupAdmissionMode._('invite-only');
  static const approvalRequired = GroupAdmissionMode._('approval');
  static const closed = GroupAdmissionMode._('closed');

  factory GroupAdmissionMode.custom(String value) {
    final trimmed = value.trim();
    if (trimmed.isEmpty) {
      throw ArgumentError.value(value, 'value', 'must not be empty');
    }
    return GroupAdmissionMode._(trimmed);
  }

  final String value;
}

class GroupMessageSecurityProfile {
  const GroupMessageSecurityProfile._(this.value);

  static const transportProtected = GroupMessageSecurityProfile._(
    'transport-protected',
  );
  static const groupE2ee = GroupMessageSecurityProfile._('group-e2ee');

  factory GroupMessageSecurityProfile.custom(String value) {
    final trimmed = value.trim();
    if (trimmed.isEmpty) {
      throw ArgumentError.value(value, 'value', 'must not be empty');
    }
    return GroupMessageSecurityProfile._(trimmed);
  }

  final String value;
}

class GroupMemberLimit {
  const GroupMemberLimit(this.value);

  final int value;
}

class CreateGroupRequest {
  const CreateGroupRequest({
    required this.name,
    this.identityMode = GroupIdentityMode.didOnly,
    this.identityHandle,
    this.description,
    this.avatarUri,
    this.discoverability,
    this.admissionMode,
    this.messageSecurityProfile,
    this.e2ee = false,
    this.slug,
    this.goal,
    this.rules,
    this.messagePrompt,
    this.docUrl,
    this.attachmentsAllowed,
    this.maxMembers,
    this.memberMaxMessages,
    this.memberMaxTotalChars,
  });

  final String name;
  final GroupIdentityMode identityMode;
  final String? identityHandle;
  final String? description;
  final String? avatarUri;
  final GroupDiscoverability? discoverability;
  final GroupAdmissionMode? admissionMode;
  final GroupMessageSecurityProfile? messageSecurityProfile;
  final bool e2ee;
  final String? slug;
  final String? goal;
  final String? rules;
  final String? messagePrompt;
  final String? docUrl;
  final bool? attachmentsAllowed;
  final GroupMemberLimit? maxMembers;
  final int? memberMaxMessages;
  final int? memberMaxTotalChars;
}

class JoinGroupRequest {
  const JoinGroupRequest({
    required this.groupDid,
    this.identityMode = GroupIdentityMode.didOnly,
    this.identityHandle,
  });

  final String groupDid;
  final GroupIdentityMode identityMode;
  final String? identityHandle;
}

class GroupRebindRecoveryItem {
  const GroupRebindRecoveryItem({
    required this.groupDid,
    required this.layer,
    required this.phase,
    required this.blocked,
  });

  final String groupDid;
  final String layer;
  final String phase;
  final bool blocked;
}

class GroupRebindRecoverySummary {
  const GroupRebindRecoverySummary({
    required this.processed,
    required this.completed,
    required this.pending,
    required this.blocked,
    this.sendPausedGroupDids = const [],
    this.items = const [],
    this.warnings = const [],
  });

  final int processed;
  final int completed;
  final int pending;
  final int blocked;
  final List<String> sendPausedGroupDids;
  final List<GroupRebindRecoveryItem> items;
  final List<String> warnings;
}

class GroupReadResult {
  const GroupReadResult({
    this.group,
    this.groups = const [],
    this.members = const [],
    required this.messages,
    this.total,
    this.source,
    this.warnings = const [],
  });

  final GroupSnapshot? group;
  final List<GroupSummary> groups;
  final List<GroupMember> members;
  final MessagePage messages;
  final int? total;
  final String? source;
  final List<String> warnings;
}

File: crates/im-core-dart/src/dto/group.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartGroupSummary {
    pub conversation_id: String,
    pub id: Option<String>,
    pub did: String,
    pub name: Option<String>,
    pub display_name: Option<String>,
    pub avatar_uri: Option<String>,
    pub my_role: Option<String>,
    pub membership_status: Option<String>,
    pub member_count: Option<u32>,
    pub last_message_at: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartGroupSnapshot {
    pub conversation_id: String,
    pub id: Option<String>,
    pub did: String,
    pub name: Option<String>,
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub avatar_uri: Option<String>,
    pub my_role: Option<String>,
    pub membership_status: Option<String>,
    pub member_count: Option<u32>,
    pub last_message_at: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartGroupMember {
    pub membership_id: Option<String>,
    pub peer_persona_id: Option<String>,
    pub did: Option<String>,
    pub credential_did: Option<String>,
    pub handle: Option<String>,
    pub role: Option<String>,
    pub status: Option<String>,
    pub joined_at: Option<String>,
    pub subject_type: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DartGroupIdentityMode {
    Handle,
    DidOnly,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartCreateGroupRequest {
    pub name: String,
    pub identity_mode: DartGroupIdentityMode,
    pub identity_handle: Option<String>,
    pub description: Option<String>,
    pub avatar_uri: Option<String>,
    pub discoverability: Option<String>,
    pub admission_mode: Option<String>,
    pub message_security_profile: Option<String>,
    pub e2ee: bool,
    pub slug: Option<String>,
    pub goal: Option<String>,
    pub rules: Option<String>,
    pub message_prompt: Option<String>,
    pub doc_url: Option<String>,
    pub attachments_allowed: Option<bool>,
    pub max_members: Option<String>,
    pub member_max_messages: Option<i64>,
    pub member_max_total_chars: Option<i64>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartJoinGroupRequest {
    pub group_did: String,
    pub identity_mode: DartGroupIdentityMode,
    pub identity_handle: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartGroupRebindRecoveryItem {
    pub group_did: String,
    pub layer: String,
    pub phase: String,
    pub blocked: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartGroupRebindRecoverySummary {
    pub processed: u32,
    pub completed: u32,
    pub pending: u32,
    pub blocked: u32,
    pub send_paused_group_dids: Vec<String>,
    pub items: Vec<DartGroupRebindRecoveryItem>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartGroupReadResult {
    pub group: Option<DartGroupSnapshot>,
    pub groups: Vec<DartGroupSummary>,
    pub members: Vec<DartGroupMember>,
    pub messages: crate::dto::message::DartMessagePage,
    pub total: Option<u32>,
    pub next_cursor: Option<String>,
    pub has_more: bool,
    pub page_group_did: Option<String>,
    pub group_state_version: Option<String>,
    pub source: Option<String>,
    pub warnings: Vec<String>,
}
